clash-prelude-hedgehog/src/Clash/Hedgehog/Sized/BitVector.hs

```diff
@@ -0,0 +1,82 @@
+{-|
+Copyright   : (C) 2021, QBayLogic B.V.
+License     : BSD2 (see the file LICENSE)
+Maintainer  : QBayLogic B.V. <devops@qbaylogic.com>
+
+Random generation of BitVector.
+-}
+
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.KnownNat.Solver #-}
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Clash.Hedgehog.Sized.BitVector
+  ( genDefinedBit
+  , genBit
+  , genDefinedBitVector
+  , genBitVector
+  , SomeBitVector(..)
+  , genSomeBitVector
+  ) where
+
+import GHC.Natural (Natural)
+import GHC.TypeNats
+import Hedgehog (MonadGen, Range)
+import qualified Hedgehog.Gen as Gen
+
+import Clash.Promoted.Nat
+import Clash.Sized.Internal.BitVector
+import Clash.Sized.Vector (v2bv)
+import Clash.XException (errorX)
+
+import Clash.Hedgehog.Sized.Vector (genVec)
+
+-- | Generate a bit which is guaranteed to be defined.
+-- This will either have the value 'low' or 'high'.
+--
+genDefinedBit :: (MonadGen m) => m Bit
+genDefinedBit = Gen.element [low, high]
+
+-- | Generate a bit which is not guaranteed to be defined.
+-- This will either have the value 'low' or 'high', or throw an 'XException'.
+--
+genBit :: (MonadGen m) => m Bit
+genBit = Gen.element [low, high, errorX "X"]
+
+-- | Generate a bit vector where all bits are defined.
+--
+genDefinedBitVector :: (MonadGen m, KnownNat n) => m (BitVector n)
+genDefinedBitVector =
+  Gen.frequency
+    [ (60, fmap v2bv (genVec genDefinedBit))
+    , (20, Gen.constant minBound)
+    , (20, Gen.constant maxBound)
+    ]
+
+-- | Generate a bit vector where some bits may be undefined.
+--
+genBitVector :: (MonadGen m, KnownNat n) => m (BitVector n)
+genBitVector =
+  Gen.frequency
+    [ (55, fmap v2bv (genVec genBit))
+    , (15, Gen.constant minBound)
+    , (15, Gen.constant maxBound)
+    , (15, Gen.constant undefined#)
+    ]
```

Comment:
This is what I'd like to see for the other generators too. I don't think it's merge blocking, but while testing I've been almost always been interested in testing extreme/interesting values (that's typically where all the bugs hid!). On the flip side, I have never been actually interested in a generator that picks values at random (after shrinking the range with some size multiplier). Still, I have used them out of sheer laziness. To safe lazy bums like me from themselves, I think our simplest generators should therefore be more likely to pick extreme values, in order to encourage usage.

Reply:
My proposal would therefore be to merge this and add an issue explaining the enhancement.